### PR TITLE
Improvements based on feedback from pull request #585

### PR DESCRIPTION
**Pull request by jthaine**

### CODE REVIEW

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 1]
```diff
@@ -2651,7 +2651,7 @@ protected void parseStubFiles() {
      * @return the annotation mirror for anno
      */
     @Override
-    public AnnotationMirror getDeclAnnotation(Element elt,
+    public final AnnotationMirror getDeclAnnotation(Element elt,
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: An `@see` to the three-argument getDeclAnnotation would be good and a note that that version should be overridden to adapt the behavior of this method and the `-NoAliases` version.

File: framework/src/org/checkerframework/framework/util/ContractsUtils.java [hunk 2]
```diff
@@ -49,12 +50,154 @@ public static ContractsUtils getInstance(
...
     /**
+     * Represents a pre- or postcondition that must be verified by
+     * {@code BaseTypeVisitor} or one of its subclasses. Automatically
+     * extracted from annotations with meta-annotations
+     * {@code @PreconditionAnnotation} or {@code @PostconditionAnnotation},
+     * such as {@code @RequiresNonNull} and {@code @EnsuresNonNull}.
+     * Can also be generated, such as in
+     * {@code LockVisitor.generatePreconditionsBasedOnGuards}.
+     */
+    public static class PreOrPostcondition {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/ContractsUtils.java]: This class is great! It made looking through the code so much easier compared to the previous Pair!

File: framework/src/org/checkerframework/framework/util/ContractsUtils.java [hunk 3]
```diff
@@ -49,12 +50,154 @@ public static ContractsUtils getInstance(
...
+
+        public PreOrPostcondition(String expression, String annotationString) {
+            this.expression = expression;
+            this.annotationString = annotationString;
+        }
+    }
+
+    /**
+     * An Iterable set of PreOrPostcondition entries.
+     * Implements the Iterable interface but not the Set interface.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/ContractsUtils.java]: I'm not so sure about the usefulness of this class and this comment. Can you document more why this is better than just using `Set<PreOrPostcondition>`?

File: framework/src/org/checkerframework/framework/util/ContractsUtils.java (hunk 2 above)
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/ContractsUtils.java]: Oh, the only thing I had wondered is if the correct camel-case is `PreOrPostCondition` to make the `Condition` part easier to notice.

File: framework/src/org/checkerframework/framework/util/ContractsUtils.java [hunk 4]
```diff
@@ -49,12 +50,154 @@ public static ContractsUtils getInstance(
...
+    }
+
+    /**
+     * Represents a conditional postcondition that must be verified by
+     * {@code BaseTypeVisitor} or one of its subclasses. Automatically
+     * extracted from annotations with meta-annotation
+     * {@code @ConditionalPostconditionAnnotation}, such as
+     * {@code EnsuresNonNullIf}.
+     */
+    public static class ConditionalPostcondition {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/ContractsUtils.java]: Should this be a subclass of `PreOrPostcondition`?
Maybe a note to keep the two classes in sync would be enough.

File: framework/src/org/checkerframework/framework/util/ContractsUtils.java [hunk 5]
```diff
@@ -49,12 +50,154 @@ public static ContractsUtils getInstance(
...
+     * Can also be generated, such as in
+     * {@code LockVisitor.generatePreconditionsBasedOnGuards}.
+     */
+    public static class PreOrPostcondition {
+        /**
+         * The expression for which the condition must hold, such as
+         * {@code "foo"} in {@code @RequiresNonNull("foo")}.
+         */
+        public final String expression;
+        /**
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/ContractsUtils.java]: Add empty line between field declarations.

**Comment by wmdietl** [comment]
Hi Javier, these are great improvements! I find the code much easier to read using the new utility classes.
Some small comments, but looks good to merge otherwise.

File: framework/src/org/checkerframework/framework/util/ContractsUtils.java (hunk 2 above)
Comment (jthaine) [inline framework/src/org/checkerframework/framework/util/ContractsUtils.java]: I think of Precondition as one word. Throughout the code we follow this convention, such as in PreconditionAnnotation and ConditionalPostconditionAnnotation.

Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/ContractsUtils.java]: True, consistency is best.

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java (hunk 1 above)
Comment (jthaine) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: The three-argument getDeclAnnotation is currently private. Should we make it protected?

Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: There is a `getDeclAnnotations(Element elt)` version which is public, but non-final.
Maybe we should make public methods which are not intended to be overridden final?
Let's just keep this as it is and see whether we need to change this later.

**Comment by wmdietl** [comment]
Final version looks really nice. Please merge.

**Comment by smillst** [comment]
Merged via 74f97e2

**Pull request closed without merging**